File: ts/test/union.test.ts
/* Copyright (c) 2024-2026 Richard Rodger, MIT License */

// An UNTAGGED union — oneOf/anyOf, two or more real branches, no
// `discriminator` — cannot be resolved to a variant by any generator: nothing
// in the schema says which branch a given value is. The field can only be
// modelled as an open type, and the point of detecting it is to let the
// generated documentation SAY why the type is open.
//
// The live case is the Typebot Builder spec, whose `groups` field is an array
// whose item schema carries 18 untagged unions, the widest 19 branches, 12
// levels down — which is why the scan has to recurse rather than look at the
// field's own schema.

import { describe, test } from 'node:test'
import { equal, deepEqual } from 'node:assert'

// Built module, matching the other suites: the compiled test runs from
// dist-test/, where a ../src path does not resolve.
import {
  untaggedUnionBranches,
  scanUntaggedUnion,
} from '../dist/utility'


describe('untagged-union', () => {

  describe('untaggedUnionBranches', () => {

    test('counts real branches of oneOf and anyOf', () => {
      equal(untaggedUnionBranches({ oneOf: [{ type: 'string' }, { type: 'number' }] }), 2)
      equal(untaggedUnionBranches(
        { anyOf: [{ type: 'string' }, { type: 'number' }, { type: 'boolean' }] }), 3)
    })


    test('a discriminated union is resolvable, so not counted', () => {
      // The discriminator names the property that decides the branch, which
      // is precisely what an untagged union lacks.
      equal(untaggedUnionBranches({
        oneOf: [{ type: 'object' }, { type: 'object' }],
        discriminator: { propertyName: 'kind' },
      }), 0)
    })


    test('the nullable idiom is one type, not a choice', () => {
      // anyOf: [X, null] means "X, possibly absent" — there is no variant to
      // pick, so flagging it would bury the real unions in noise.
      equal(untaggedUnionBranches({ anyOf: [{ type: 'string' }, { type: 'null' }] }), 0)
      equal(untaggedUnionBranches({ oneOf: [{ type: 'object' }, { type: 'null' }] }), 0)
    })


    test('ignores non-unions', () => {
      equal(untaggedUnionBranches({ type: 'string' }), 0)
      equal(untaggedUnionBranches({ allOf: [{ type: 'object' }, { type: 'object' }] }), 0)
      equal(untaggedUnionBranches({ oneOf: [{ type: 'string' }] }), 0)
      equal(untaggedUnionBranches({ oneOf: [] }), 0)
      equal(untaggedUnionBranches(null), 0)
      equal(untaggedUnionBranches('nope'), 0)
    })

  })


  describe('scanUntaggedUnion', () => {

    test('null when nothing beneath the field is a union', () => {
      equal(scanUntaggedUnion({ type: 'object', properties: { a: { type: 'string' } } }), null)
      equal(scanUntaggedUnion(null), null)
    })


    test('finds a union at the field itself', () => {
      deepEqual(
        scanUntaggedUnion({ oneOf: [{ type: 'string' }, { type: 'number' }] }),
        { count: 1, branches: 2, depth: 0 })
    })


    test('finds a union nested below the field, reporting its depth', () => {
      // The shape of the Typebot `groups` field: an array whose items carry
      // the union.
      const schema = {
        type: 'array',
        items: {
          type: 'object',
          properties: {
            blocks: {
              type: 'array',
              items: { anyOf: [{ type: 'object' }, { type: 'object' }, { type: 'object' }] },
            },
          },
        },
      }
      const found = scanUntaggedUnion(schema)
      equal(found?.count, 1)
      equal(found?.branches, 3)
      // The walk is generic over object values, so the `properties` container
      // counts as a level of its own: schema -> items -> properties -> blocks
      // -> items. Depth is a relative "how far down", not a JSON-Pointer hop
      // count, and is only ever compared against other depths.
      equal(found?.depth, 4)
    })


    test('reports the WIDEST union and the total count', () => {
      const schema = {
        a: { oneOf: [{ type: 'string' }, { type: 'number' }] },
        b: { anyOf: [{ type: 'a' }, { type: 'b' }, { type: 'c' }, { type: 'd' }] },
      }
      const found = scanUntaggedUnion(schema)
      equal(found?.count, 2)
      equal(found?.branches, 4)
    })


    test('survives a self-referential schema', () => {
      // These specs reference themselves freely; an unguarded walk would spin.
      const node: any = { type: 'object', properties: {} }
      node.properties.self = node
      node.properties.choice = { oneOf: [{ type: 'string' }, { type: 'number' }] }
      const found = scanUntaggedUnion(node)
      equal(found?.count, 1)
      equal(found?.branches, 2)
    })

  })

})

File: go/union_test.go
package apidef

import "testing"

// Parity port of ts/test/union.test.ts. An UNTAGGED union — oneOf/anyOf, two
// or more real branches, no discriminator — cannot be resolved to a variant by
// any generator, so the field can only be modelled as an open type.

func TestUntaggedUnionBranches(t *testing.T) {
	cases := []struct {
		name   string
		schema map[string]any
		want   int
	}{
		{
			name:   "oneOf two branches",
			schema: map[string]any{"oneOf": []any{map[string]any{"type": "string"}, map[string]any{"type": "number"}}},
			want:   2,
		},
		{
			name: "anyOf three branches",
			schema: map[string]any{"anyOf": []any{
				map[string]any{"type": "string"},
				map[string]any{"type": "number"},
				map[string]any{"type": "boolean"},
			}},
			want: 3,
		},
		{
			// The discriminator names the property that decides the branch,
			// which is exactly what an untagged union lacks.
			name: "discriminated union is resolvable",
			schema: map[string]any{
				"oneOf":         []any{map[string]any{"type": "object"}, map[string]any{"type": "object"}},
				"discriminator": map[string]any{"propertyName": "kind"},
			},
			want: 0,
		},
		{
			// anyOf: [X, null] is "X, possibly absent" — no variant to pick.
			name:   "nullable idiom is not a union",
			schema: map[string]any{"anyOf": []any{map[string]any{"type": "string"}, map[string]any{"type": "null"}}},
			want:   0,
		},
		{
			name:   "allOf is composition, not choice",
			schema: map[string]any{"allOf": []any{map[string]any{"type": "object"}, map[string]any{"type": "object"}}},
			want:   0,
		},
		{
			name:   "single branch",
			schema: map[string]any{"oneOf": []any{map[string]any{"type": "string"}}},
			want:   0,
		},
		{
			name:   "plain scalar",
			schema: map[string]any{"type": "string"},
			want:   0,
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := UntaggedUnionBranches(tc.schema); got != tc.want {
				t.Fatalf("UntaggedUnionBranches = %d, want %d", got, tc.want)
			}
		})
	}
}

func TestScanUntaggedUnionNil(t *testing.T) {
	if got := ScanUntaggedUnion(map[string]any{
		"type":       "object",
		"properties": map[string]any{"a": map[string]any{"type": "string"}},
	}); got != nil {
		t.Fatalf("expected nil for a resolvable schema, got %+v", got)
	}
	if got := ScanUntaggedUnion(nil); got != nil {
		t.Fatalf("expected nil for nil schema, got %+v", got)
	}
}

func TestScanUntaggedUnionAtField(t *testing.T) {
	got := ScanUntaggedUnion(map[string]any{
		"oneOf": []any{map[string]any{"type": "string"}, map[string]any{"type": "number"}},
	})
	if got == nil {
		t.Fatal("expected a union")
	}
	if got.Count != 1 || got.Branches != 2 || got.Depth != 0 {
		t.Fatalf("got %+v, want {Count:1 Branches:2 Depth:0}", got)
	}
}

func TestScanUntaggedUnionNested(t *testing.T) {
	// The shape of the Typebot `groups` field: an array whose items carry the
	// union, several levels below the field itself.
	schema := map[string]any{
		"type": "array",
		"items": map[string]any{
			"type": "object",
			"properties": map[string]any{
				"blocks": map[string]any{
					"type": "array",
					"items": map[string]any{"anyOf": []any{
						map[string]any{"type": "object"},
						map[string]any{"type": "object"},
						map[string]any{"type": "object"},
					}},
				},
			},
		},
	}
	got := ScanUntaggedUnion(schema)
	if got == nil {
		t.Fatal("expected a union")
	}
	if got.Count != 1 || got.Branches != 3 {
		t.Fatalf("got %+v, want Count:1 Branches:3", got)
	}
	if got.Depth == 0 {
		t.Fatal("expected a non-zero depth for a nested union")
	}
}

func TestScanUntaggedUnionWidest(t *testing.T) {
	got := ScanUntaggedUnion(map[string]any{
		"a": map[string]any{"oneOf": []any{
			map[string]any{"type": "string"}, map[string]any{"type": "number"}}},
		"b": map[string]any{"anyOf": []any{
			map[string]any{"type": "a"}, map[string]any{"type": "b"},
			map[string]any{"type": "c"}, map[string]any{"type": "d"}}},
	})
	if got == nil {
		t.Fatal("expected a union")
	}
	if got.Count != 2 || got.Branches != 4 {
		t.Fatalf("got %+v, want Count:2 Branches:4", got)
	}
}

func TestScanUntaggedUnionCyclic(t *testing.T) {
	// These specs reference themselves freely; an unguarded walk would spin.
	node := map[string]any{"type": "object"}
	props := map[string]any{
		"self": node,
		"choice": map[string]any{"oneOf": []any{
			map[string]any{"type": "string"}, map[string]any{"type": "number"}}},
	}
	node["properties"] = props

	got := ScanUntaggedUnion(node)
	if got == nil {
		t.Fatal("expected a union")
	}
	if got.Count != 1 || got.Branches != 2 {
		t.Fatalf("got %+v, want Count:1 Branches:2", got)
	}
}
